server/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -134,6 +135,61 @@ public StoredContext stashContext() {
+     * value (including restoring them to an <i>unset</i> value if they did not originally exist).
+     * Closing the {@code StoredContext} has no affect on any other header - any headers
+     * (other than those names specified in {@code transientHeadersToStash} that were
+     * added to the {@code ThreadContext} will be retained.
+     *
+     * For example, at the end of the following code, the ThreadContext will have transient
+     * values {@code "a"=1}, {@code "b"=1}, {@code "d"=2} and {@code "c"} will not be set.
+     * <pre>
+     * threadContext.putTransient("a", 1);
+     * threadContext.putTransient("b", 1);
+     * try (ThreadContext.StoredContext restore = threadContext.stashTransientContext(List.of("b", "c")) ) {
+     *   threadContext.putTransient("b", 2);
+     *   threadContext.putTransient("c", 2);
+     *   threadContext.putTransient("d", 2);
+     * }
+     * </pre>
+     */
+    public StoredContext stashTransientHeaders(Collection<String> transientHeadersToStash) {
+        if (transientHeadersToStash.isEmpty()) {
+            // no-op
+            return () -> {
+            };
+        }
+        final ThreadContextStruct beforeContext = threadLocal.get();
+        final Map<String, Object> stashedTransientHeaders = new HashMap<>();
+        Map<String, Object> newBeforeTransientHeaders = new HashMap<>(beforeContext.transientHeaders);
+        for (String transientHeaderToStash : transientHeadersToStash) {
+            if (newBeforeTransientHeaders.containsKey(transientHeaderToStash)) {
+                Object stashed = newBeforeTransientHeaders.remove(transientHeaderToStash);
+                stashedTransientHeaders.put(transientHeaderToStash, stashed);
+            }
+        }
+        if (false == stashedTransientHeaders.isEmpty()) {
+            ThreadContextStruct threadContextStruct = new ThreadContextStruct(beforeContext.requestHeaders, beforeContext.responseHeaders,
+                    newBeforeTransientHeaders, beforeContext.isSystemContext);
+            threadLocal.set(threadContextStruct);
+        }
+        return () -> {
+            final ThreadContextStruct afterContext = threadLocal.get();
+            Map<String, Object> newAfterTransientHeaders = new HashMap<>(afterContext.transientHeaders);
+            // remove the transients that might have been set in the mean time
+            for (String transientHeaderToStash : transientHeadersToStash) {
+                newAfterTransientHeaders.remove(transientHeaderToStash);
+            }
+            // fill in the values that existed before stashing
+            newAfterTransientHeaders.putAll(stashedTransientHeaders);
+            ThreadContextStruct threadContextStruct = new ThreadContextStruct(afterContext.requestHeaders, afterContext.responseHeaders,
+                    newAfterTransientHeaders, afterContext.isSystemContext);
+            threadLocal.set(threadContextStruct);
```

Comment:
Why can't we just restore the beforeContext?

```suggestion
            return () -> threadLocal.set(beforeContext);
```

Reply:
We have to maintain all the other headers (transient, request, response) as they might have been modified inside the stash, and revert only the ones in the arg list.